code_generator/GdkCodeGenerator/Templates/UnityGameObjectComponentDispatcherGenerator.tt

```diff
@@ -20,58 +20,106 @@ namespace <#= qualifiedNamespace #>
             {
 <# foreach(var eventDetails in eventDetailsList) { #>
-                new ComponentType[] { typeof(EventsReceived<<#= eventDetails.EventName #>Event>), typeof(GameObjectReference) },
+                new ComponentType[] { ComponentType.ReadOnly<EventsReceived<<#= eventDetails.EventName #>Event>>(), ComponentType.ReadOnly<GameObjectReference>() },
 <# } #>
             };
 
             public override ComponentType[][] CommandRequestsComponentTypeArrays => new ComponentType[][]
             {
 <# foreach(var commandDetails in commandDetailsList) { #>
-                new ComponentType[] { typeof(CommandRequests<<#= commandDetails.CommandName #>.Request>), typeof(GameObjectReference) },
+                new ComponentType[] { ComponentType.ReadOnly<CommandRequests<<#= commandDetails.CommandName #>.Request>>(), ComponentType.ReadOnly<GameObjectReference>() },
 <# } #>
             };
 
-            public override void InvokeOnAddComponentLifecycleCallbacks(GameObjectDispatcherSystem gameObjectDispatcherSystem)
+            protected override uint getComponentId()
             {
-                var entities = ComponentAddedComponentGroup.GetEntityArray();
+                return <#= unityComponentDefinition.Id #>;
+            }
+
+            public override void InvokeOnComponentUpdateUserCallbacks(GameObjectDispatcherSystem gameObjectDispatcherSystem)
+            {
+<# if (fieldDetailsList.Count > 0) { #>
+                var entities = ComponentsUpdatedComponentGroup.GetEntityArray();
+                var updateLists = ComponentsUpdatedComponentGroup.GetComponentArray<ComponentsUpdated<<#= componentDetails.TypeName #>.Update>();
                 for (var i = 0; i < entities.Length; i++)
                 {
-                    var spatialOSBehaviourManager = gameObjectDispatcherSystem.GetSpatialOSBehaviourManager(entities[i].Index);
-                    spatialOSBehaviourManager.AddComponent(<#= unityComponentDefinition.Id #>);
+                    var readers = gameObjectDispatcherSystem.GetSpatialOSBehaviourManager(entities[i].Index)
+                        .GetReadersWriters(getComponentId());
+                    var updateList = updateLists[i];
+                    foreach (var reader in readers)
+                    {
+                        foreach (var update in updateList.Buffer)
+                        {
+                            ((<#= componentDetails.TypeName #>.ReaderWriterImpl) reader).OnComponentUpdate(update);
+                        }
+                    }
                 }
+<# } #>
             }
 
-            public override void InvokeOnRemoveComponentLifecycleCallbacks(GameObjectDispatcherSystem gameObjectDispatcherSystem)
+            public override void InvokeOnEventUserCallbacks(GameObjectDispatcherSystem gameObjectDispatcherSystem)
             {
-                var entities = ComponentRemovedComponentGroup.GetEntityArray();
+<# for (var i = 0; i < eventDetailsList.Count; i++) { #>
```

Comment:
Might be worth using `j` instead of `i` here given that the templated code below also uses `i` as an index